### DISCUSSION TITLE
Reset `FingersCrossedHandler` after request

### DISCUSSION THREAD

**Discussion by CViniciusSDias** [symfony/symfony#54595] [Q&A]
After a message is handled, symfony resets the container, which can reset the `FingersCrossedHandler` from monolog.
That should probably also be done after requests are processed in statefull runtimes (such as Swoole), right?

**Comment by ghost** [comment]
Symfony doesn't know your runtime is Swoole. You can create your own implementation of \Symfony\Component\Runtime\RuntimeInterface  for Swoole that will call ServicesResetter or register an event listener that resets services on kernel.terminate (with the lowest priority possible).

**Reply by CViniciusSDias** [reply]
But Symfony doesn't need to know that the runtime is Swoole. That's my point. I believe the FingersCrossedHandler should always be reset on terminate. If it's running on a stateless runtime, there's no harm. If it's in a stateful runtime (whichever it is), it's basically necessary.

**Reply by ghost** [reply]
> If it's running on a stateless runtime, there's no harm

Yes there is. Resetting services might be a realitively heavy task. Ofc one could say that `kernel.terminate` event is dispatched after the response has been sent so the ovehead does not affect user (this is true for FPM, might not be a case for Swoole). Maybe.
To sum up, the decision whether to reset services at the end (or at the beginning) of the request or not can be left to the developer itself. Just add the event listener, it's really simple.

**Reply by CViniciusSDias** [reply]
I don't believe it's simple since not everyone knows about FingersCrossedHandler and since that's the default in prod, I believe a safer approach should also be the default one for the framework. But I do get your point.

**Reply by stof** [reply]
MonologBundle should already be wiring the FingersCrossedHandler as a resettable service in the ServicesResetter. When running in long-running runtimes, you are expected to use this ServicesResetter. But doing that for everyone, even when using short-lived runtimes (which is the 99% use case) would be bad for performance.

**Reply by CViniciusSDias** [reply]
I brought this up [here](https://github.com/Seldaek/monolog/issues/1885) in case anyone wants to participate in the discussion.

**Reply by CViniciusSDias** [reply]
> MonologBundle should already be wiring the FingersCrossedHandler as a resettable service in the ServicesResetter.

Interesting. Where do I find documentation about this `ServicesResetter` or how do I see if that's already wired?

**Reply by ghost** [reply]
Check whether FingersCrossedHandler has `kernel.reset` tag with `bin/console debug:container --tag=kernel.reset` (if you use FingersCrossedHandler only in production then add `-e prod`)

**Reply by CViniciusSDias** [reply]
Yep. It is marked as resettable. Thank you.
But where do I find documentation on that matter? I didn't find anything on "ServicesResetter" in Symfony docs.

**Reply by ghost** [reply]
`kernel.reset` tag is documented here https://symfony.com/doc/current/reference/dic_tags.html#kernel-reset. The class ServicesResetter is not documented because it's an impementation detail.

**Reply by ghost** [reply]
Btw. I can see the documentation says:
> In all main requests (not [sub-requests](https://symfony.com/doc/current/components/http_kernel.html#http-kernel-sub-requests)) except the first one, Symfony looks for any service tagged with the kernel.reset tag to reinitialize their state.

That's interesting. If I understand correctly Symfony resets all the services registered with this tag on every next request out of the box. I didn't know that.  Will have to verify this myself in the spare time.
